### DISCUSSION TITLE
🔥Out now: improved support for fixed-format RPGLE 🔥

### DISCUSSION THREAD

**Discussion by worksofliam** [codefori/vscode-ibmi#2406] [Announcements]
It brings me great pleasure to announce that for the RPGLE extension, 0.28.0, improvements to the RPGLE language server will support fixed-format RPGLE. Most importantly, we are shipping the ability to use the _Peek_ and _Go to References_ within Visual Studio Code.

This has been a long standing issue for people transitioning to Visual Studio Code. I am hoping this is one of the bigger technical hurdles that will help people as they start using it.

* Check out our docs on [RPGLE in Visual Studio Code](https://codefori.github.io/docs/extensions/rpgle/) ✍️
* Make sure to have the [RPGLE extension from the Marketplace](https://marketplace.visualstudio.com/items?itemName=HalcyonTechLtd.vscode-rpgle) 👩‍💻

I am expecting 0.28.0 to come out for 2025.

---

Here is an example of the Peek view that will be available:

<img width="1365" alt="image" src="https://github.com/user-attachments/assets/ff855f42-6149-4c81-99a8-d09398ebca03">

And of course, there is the standard References view also:

<img width="1526" alt="image" src="https://github.com/user-attachments/assets/30b6e4c4-6a76-432b-bad0-acd9e74c568c">

And as a reminder, here is _Peek Definition_ for variables:

<img width="1526" alt="image" src="https://github.com/user-attachments/assets/efa618c7-605f-4287-9061-f9cf636c5cdf">

Here is a bonus feature - we're adding the column outline for the current line you're editing on!

![image](https://github.com/user-attachments/assets/6d3b9fe5-023f-4f7b-b205-612e13246a59)

More new feature include being able to use _go to implementation_ to find where a procedure is defined (not the prototype, the actual procedure):

![telegram-cloud-photo-size-1-5100642928564481603-y](https://github.com/user-attachments/assets/07d0ccb1-f5fc-49ef-9260-1707d377e3e7)

**Comment by jkyeung** [comment]
This is a big deal. Most of us are mainly maintaining existing code, which naturally is mostly in fixed format. For long-time SEU users who have never tried RDi or other full-featured IDE, this is an eye-popping, "wow" feature. It will go a long way toward enticing old-schoolers to adopt VS Code. And of course, it significantly helps existing VS Code users work with their legacy RPG codebases.

**Reply by JDubb-At-Nelnet** [reply]
love the column outline!

**Comment by rogersonra** [comment]
It's great all the enhancements that you guys are doing.

That said, is there a way to turn off the column outline?  

My problem is that it hides the previous line.  Sometimes that line is important to the logic.  
For example, the AND line is being hidden behind the outline.
![image](https://github.com/user-attachments/assets/30dc0b60-02ef-49c1-a4f2-ea40e54a7748)
![image](https://github.com/user-attachments/assets/44bebf7c-fd06-4279-a7d5-3c12b61eca20)

Thanks,

Rob

**Reply by jkyeung** [reply]
Not sure if this is technically feasible, but even better still would be for the column outline to be on its own, unnumbered line.

Assuming it is possible to turn off completely, making it easy to toggle on and off would be preferable to having a configuration setting that enables or disables it.

What does it look like in SEU? (I actually never see this myself, and don't know how to make it show up. A colleague reports that when she's in SEU, it shows up and goes away kind of randomly, and she doesn't know how to control it or configure it.)

**Reply by worksofliam** [reply]
@rogersonra yes, it can be turned off in the VS Code settings under RPGLE.

**Reply by rogersonra** [reply]
In order to turn off the column outline, open Extensions, RPGLE and open the Settings.
![image](https://github.com/user-attachments/assets/36fff921-c76c-4c2f-aada-8ea23b0112e6)

Uncheck the Whether to show the fixed-format ruler by default.
![image](https://github.com/user-attachments/assets/7206e957-e45a-48d3-a9d4-ef331ecf1a6c)

**Reply by jkyeung** [reply]
@rogersonra - Thanks for that information.

For completeness, I'll mention that [Shift+F4](https://codefori.github.io/docs/extensions/rpgle/#fixed-format-ruler) is the way to enable or disable the ruler on demand.

Small quirk: If you've just toggled the ruler off, and immediately try to toggle it back on, it won't show up until you move the cursor to a different line.

<sub>_More precisely: When you try to toggle the ruler back on, if the cursor is on the same line as where you toggled it off, the ruler won't show up. There can be intervening cursor movement and even edits between toggling off and back on, but if you put the cursor back on the toggle-off line just before toggling on, then the ruler won't show up until you move the cursor to some other line._</sub>

**Reply by NextPage66** [reply]
Just making sure I am reading this correctly.
You are using Shift-F4 to enable/disable the ruler....correct?
That sequence of keys is opening the FIND feature in my source....not toggling the ruler.

**Reply by jkyeung** [reply]
@NextPage66 - Correct. But you may have different keybindings, especially if you have different extensions. You can use the Keyboard Shortcuts dialog to find what commands are attached to a particular key combo, and edit as necessary/desired.

**Reply by NextPage66** [reply]
Here is my keybinding.  I am in an rpgle source file member and the system language mode is designated as RPGLE.  It is still not working.  What am I missing?
![image](https://github.com/user-attachments/assets/1acd8527-bbca-4267-b6ac-80a3d749e9cb)

**Reply by jkyeung** [reply]
Looks to me as though the `editorFocus` condition takes precedence over the `editorLangId == 'rpgle'` condition. I would try editing the ones coming from your Notepad++ extension so that their condition is `editorFocus && editorLangId != 'rpgle'`.

**Reply by NextPage66** [reply]
I can see the key bindings but not allowed to change the WHEN condition.  Can only change the keys used.
Is this list of key bindings from the Microsoft system or is it self contained in VSCode?

**Reply by DamianoBonato** [reply]
VSCode

**Reply by DamianoBonato** [reply]
Resolved, there was a command conflict .
Thanks .

**Comment by DamianoBonato** [comment]
Good evening, when I press Alt+F4 the following message appears:
![image](https://github.com/user-attachments/assets/ba65bf1d-5311-4632-8e82-cc420be93714)
